xcube_zenodo/preload.py

```diff
@@ -0,0 +1,298 @@
+# The MIT License (MIT)
+# Copyright (c) 2024 by the xcube development team and contributors
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import shutil
+import threading
+import time
+from typing import Callable
+import tarfile
+import zipfile
+
+import fsspec
+import numpy as np
+import tabulate
+import requests
+import xarray as xr
+from xcube.core.store import DataStoreError
+from xcube.core.store import MutableDataStore
+
+from ._utils import identify_file_format
+from ._utils import translate_data_id2uri
+from .constants import LOG
+
+
+class Event:
+
+    def __init__(self, data_id: str, total_size: int | float):
+        self.data_id = data_id
+        self.status = "Not started"
+        self.progress = 0.0
+        self.message = "Preloading not started yet."
+        self.total_size = total_size
+        self._callback = None
+
+    def subscribe(self, callback: Callable[[], None]):
+        self._callback = callback
+
+    def notify(self):
+        if self._callback is not None:
+            self._callback()
+
+    def update(self, status: str, progress: float, message: str):
+        self.status = status
+        self.progress = progress
+        self.message = message
+        self.notify()
+
+
+class PreloadHandle:
+
+    def __init__(self, cache_store: MutableDataStore, *data_ids: str, **preload_params):
+        self._is_cancelled = False
+        self._is_closed = False
+        self._cache_store = cache_store
+        self._cache_fs: fsspec.AbstractFileSystem = cache_store.fs
```

Comment:
@forman can give me feedback on this. I removed `os` completely, but I readout `root` and `fs` from the cache datastore, which I doubt is correct.

Reply:
No, that's fine, `root` and `fs` belong to the `FsDataStore`'s API.